consensus/istanbul/core/roundchange_test.go

```diff
@@ -91,3 +92,371 @@ func TestRoundChangeSet(t *testing.T) {
+		t.Error("Did not finalize a block in round 1")
+	case _, ok := <-newBlocks.Chan():
+		if !ok {
+			t.Error("Error reading block")
+		}
+		// Wait for all backends to finalize the block.
+		<-time.After(1 * time.Second)
+		for i, b := range sys.backends {
+			committed, _ := b.LastProposal()
+			// We expect to commit the block proposed by the first proposer.
+			expectedCommitted := makeBlockWithDifficulty(1, 0)
+			if expectedCommitted.Hash() != committed.Hash() {
+				t.Errorf("Backend %v got committed block with unexpected hash: expected %v, got %v", i, expectedCommitted.Hash(), committed.Hash())
+			}
+		}
+	}
+
+	// Manually open and close b/c hijacking sys.listen
+	for _, b := range sys.backends {
+		b.engine.Stop() // start Istanbul core
+	}
+	close(sys.quit)
+}
+
+// This tests that when F+1 nodes receive 2F+1 PREPARE messages for a particular proposal, the
+// system enforces that as the only valid proposal for this sequence.
+func TestPreparedCertificatePersistsThroughRoundChanges(t *testing.T) {
+	// Initialize the system with a nil round state so that we properly start round 0.
+	sys := NewTestSystemWithBackendAndCurrentRoundState(4, 1, func(vset istanbul.ValidatorSet) *roundState { return nil })
+
+	for i, b := range sys.backends {
+		b.engine.Start() // start Istanbul core
+		block := makeBlockWithDifficulty(1, int64(i))
+		sys.backends[i].NewRequest(block)
+	}
+
+	newBlocks := sys.backends[3].EventMux().Subscribe(istanbul.FinalCommittedEvent{})
+	defer newBlocks.Unsubscribe()
+
+	timeout := sys.backends[3].EventMux().Subscribe(timeoutEvent{})
+	defer timeout.Unsubscribe()
+
+	istMsgDistribution := map[uint64]map[int]bool{}
+
+	// Send PREPARE messages to F + 1 nodes so we guarantee a PREPARED certificate in the ROUND CHANGE certificate..
+	istMsgDistribution[istanbul.MsgPreprepare] = gossip
+	istMsgDistribution[istanbul.MsgPrepare] = sendToFPlus1
+	istMsgDistribution[istanbul.MsgCommit] = gossip
+	istMsgDistribution[istanbul.MsgRoundChange] = gossip
+
+	go sys.distributeIstMsgs(t, sys, istMsgDistribution)
+
+	// Turn PREPARE messages off for round 1 to force reuse of the PREPARED certificate.
+	<-time.After(1 * time.Second)
+	istMsgDistribution[istanbul.MsgPrepare] = noGossip
+
+	// Wait for round 1 to start.
+	<-timeout.Chan()
+	// Turn PREPARE messages back on in time for round 2.
+	<-time.After(1 * time.Second)
+	istMsgDistribution[istanbul.MsgPrepare] = gossip
+
+	// Wait for round 2 to start.
+	<-timeout.Chan()
+
+	select {
+	case <-timeout.Chan():
+		t.Error("Did not finalize a block in round 2.")
+	case _, ok := <-newBlocks.Chan():
+		if !ok {
+			t.Error("Error reading block")
+		}
+		// Wait for all backends to finalize the block.
+		<-time.After(2 * time.Second)
+		for i, b := range sys.backends {
+			// TODO(asa): Do something with proposer here
+			committed, _ := b.LastProposal()
+			// We expect to commit the block proposed by the first proposer.
+			// TODO(joshua): This can be flaky in my local tests.
```

Comment:
Any insights into this?

Reply:
Unfortunately not. I just saw it be flaky.